File: app/src/main/java/com/esorokin/lantector/ui/activity/BaseActivity.java
package com.esorokin.lantector.ui.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.MenuItem;

import com.arellomobile.mvp.MvpAppCompatActivity;
import com.esorokin.lantector.ui.plugins.base.CompositionPlugin;
import com.esorokin.lantector.ui.plugins.base.CompositionPluginDelegate;

public class BaseActivity extends MvpAppCompatActivity {
	private final CompositionPlugin compositionPlugin = new CompositionPluginDelegate();

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		initPlugins();
		compositionPlugin.onCreate();
	}

	/**
	 * Attach your plugins here.
	 */
	protected void initPlugins() {
		//override
	}

	protected CompositionPlugin compositionPlugin() {
		return compositionPlugin;
	}

	@Override
	protected void onStart() {
		super.onStart();
		compositionPlugin.onStart();
	}

	@Override
	protected void onResume() {
		super.onResume();
		compositionPlugin.onResume();
	}

	@Override
	public void onContentChanged() {
		super.onContentChanged();
		compositionPlugin.onViewCreated(getWindow().getDecorView());
	}

	@Override
	protected void onPause() {
		super.onPause();
		compositionPlugin.onPause();
	}

	@Override
	protected void onStop() {
		super.onStop();
		compositionPlugin.onStop();
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		compositionPlugin.onDestroy();
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		compositionPlugin.onOptionsItemSelected(item);
		return super.onOptionsItemSelected(item);
	}

	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
		compositionPlugin.onActivityResult(requestCode, resultCode, data);
	}

	@Override
	public void onBackPressed() {
		super.onBackPressed();
		compositionPlugin.onBackPressed();
	}

	@Override
	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
		super.onRequestPermissionsResult(requestCode, permissions, grantResults);
		compositionPlugin.onRequestPermissionsResult(requestCode, permissions, grantResults);
	}
}

File: app/src/main/java/com/esorokin/lantector/ui/activity/BaseActivity.kt
package com.esorokin.lantector.ui.activity

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import com.arellomobile.mvp.MvpAppCompatActivity
import com.esorokin.lantector.ui.plugin.MoxyActivityPlugin
import com.esorokin.lantector.ui.plugin.base.CompositionPluginDelegate

abstract class BaseActivity : AppCompatActivity() {
    protected val compositionPlugin = CompositionPluginDelegate()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initPlugins()
        compositionPlugin.onCreate(savedInstanceState)
    }

    /**
     * Attach your plugins here.
     */
    protected open fun initPlugins() {
        compositionPlugin.attach(MoxyActivityPlugin(this))
    }

    override fun onStart() {
        super.onStart()
        compositionPlugin.onStart()
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
        super.onRestoreInstanceState(savedInstanceState)
        compositionPlugin.onRestoreInstanceState(savedInstanceState)
    }

    override fun onResume() {
        super.onResume()
        compositionPlugin.onResume()
    }

    override fun onContentChanged() {
        super.onContentChanged()
        compositionPlugin.onViewCreated(window.decorView)
    }

    override fun onPause() {
        super.onPause()
        compositionPlugin.onPause()
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        compositionPlugin.onSaveInstanceState(outState)
    }

    override fun onStop() {
        super.onStop()
        compositionPlugin.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        compositionPlugin.onDestroy()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        compositionPlugin.onOptionsItemSelected(item)
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        compositionPlugin.onActivityResult(requestCode, resultCode, data)
    }

    override fun onBackPressed() {
        super.onBackPressed()
        compositionPlugin.onBackPressed()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        compositionPlugin.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }
}
